cdc/redo/writer/memory/file_worker.go

```diff
@@ -0,0 +1,290 @@
+		extStorage:    extStorage,
+		uuidGenerator: uuid.NewGenerator(),
+		pool: sync.Pool{
+			New: func() interface{} {
+				// Use pointer here to prevent static checkers from reporting errors.
+				// Ref: https://github.com/dominikh/go-tools/issues/1336.
+				buf := make([]byte, 0, cfg.MaxLogSizeInBytes)
+				return &buf
+			},
+		},
+		flushCh: make(chan *fileCache),
+		metricWriteBytes: common.RedoWriteBytesGauge.
+			WithLabelValues(cfg.ChangeFeedID.Namespace, cfg.ChangeFeedID.ID),
+		metricFlushAllDuration: common.RedoFlushAllDurationHistogram.
+			WithLabelValues(cfg.ChangeFeedID.Namespace, cfg.ChangeFeedID.ID),
+	}
+}
+
+func (f *fileWorkerGroup) Run(
+	ctx context.Context, inputCh <-chan *polymorphicRedoEvent,
+) (err error) {
+	defer func() {
+		f.close()
+		if err != nil {
+			log.Warn("redo file workers closed with error", zap.Error(err))
+		}
+	}()
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return f.bgWriteLogs(egCtx, inputCh)
+	})
+	for i := 0; i < f.workerNum; i++ {
+		eg.Go(func() error {
+			return f.bgFlushFileCache(egCtx)
+		})
+	}
+	log.Info("redo file workers started", zap.Int("workerNum", f.workerNum))
+	return eg.Wait()
+}
+
+func (f *fileWorkerGroup) close() {
+	common.RedoFlushAllDurationHistogram.
+		DeleteLabelValues(f.cfg.ChangeFeedID.Namespace, f.cfg.ChangeFeedID.ID)
+	common.RedoWriteBytesGauge.
+		DeleteLabelValues(f.cfg.ChangeFeedID.Namespace, f.cfg.ChangeFeedID.ID)
+}
+
+func (f *fileWorkerGroup) bgFlushFileCache(egCtx context.Context) error {
+	for {
+		select {
+		case <-egCtx.Done():
+			return errors.Trace(egCtx.Err())
+		case file := <-f.flushCh:
+			start := time.Now()
+			err := f.extStorage.WriteFile(egCtx, file.filename, file.data)
+			f.metricFlushAllDuration.Observe(time.Since(start).Seconds())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			file.markFlushed()
+
+			bufPtr := &file.data
+			file.data = nil
+			f.pool.Put(bufPtr)
+		}
+	}
+}
+
+func (f *fileWorkerGroup) bgWriteLogs(
+	egCtx context.Context, inputCh <-chan *polymorphicRedoEvent,
+) (err error) {
+	for {
+		select {
+		case <-egCtx.Done():
+			return errors.Trace(egCtx.Err())
+		case event := <-inputCh:
+			if event == nil {
+				log.Panic("inputCh of redo file worker is closed unexpectedly")
+			}
+
+			if event.data != nil {
+				err = f.writeToCache(egCtx, event)
+				event.reset()
+				eventPool.Put(event)
+			} else if event.flushCallback != nil {
+				err = f.flushAll(egCtx)
+				event.flushCallback()
+			}
+
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+// newFileCache write event to a new file cache.
+func (f *fileWorkerGroup) newFileCache(event *polymorphicRedoEvent) {
+	bufPtr := f.pool.Get().(*[]byte)
+	file := newFileCache(event, *bufPtr)
+	f.files = append(f.files, file)
+}
+
+func (f *fileWorkerGroup) writeToCache(
+	egCtx context.Context, event *polymorphicRedoEvent,
+) error {
+	writeLen := int64(event.data.Len())
+	if writeLen > f.cfg.MaxLogSizeInBytes {
```

Comment:
What if we encounter a large row and the encoded data length exceeds `MaxLogSizeInBytes`?

Reply:
In this case, an error will be returned to the caller.